worker/agent.go

```diff
@@ -32,143 +39,216 @@ func newAgent(net, addr string, worker *Worker) (a *agent, err error) {
+	for startConnVersion == a.connectionVersion && !a.worker.isShuttingDown() {
+
+		if data, err = a.read(); err != nil {
+			if opErr, ok := err.(*net.OpError); ok {
+				if opErr.Temporary() {
+					a.worker.Log(Info, "opErr.Temporary():", a.addr)
+					continue
+				} else {
+					a.worker.Log(Info, "got permanent network error with server:", a.addr, "comm thread exiting.")
+					a.reconnect_error(err)
+					// else - we're probably dc'ing due to a Close()
 					break
 				}
-				a.rw = bufio.NewReadWriter(bufio.NewReader(a.conn),
-					bufio.NewWriter(a.conn))
-			}
-			if len(leftdata) > 0 { // some data left for processing
-				data = append(leftdata, data...)
+			} else {
+				a.worker.Log(Info, "got error", err.Error(), "with server:", a.addr, "comm thread exiting...")
+				a.reconnect_error(err)
+				break
 			}
-			if len(data) < rt.MinPacketLength { // not enough data
-				leftdata = data
-				continue
-			}
-			for {
-				if inpack, l, err = decodeInPack(data); err != nil {
-					a.worker.err(err)
-					leftdata = data
+		}
+		if len(leftdata) > 0 { // some data left for processing
+			data = append(leftdata, data...)
+		}
+		if len(data) < rt.MinPacketLength { // not enough data
+			leftdata = data
+			continue
+		}
+		for {
+			if inpack, l, err = decodeInPack(data); err != nil {
+				a.reconnect_error(err)
+				break
+			} else {
+				leftdata = nil
+				inpack.a = a
+				a.worker.in <- inpack
+				if len(data) == l {
 					break
-				} else {
-					leftdata = nil
-					inpack.a = a
-					a.worker.in <- inpack
-					if len(data) == l {
-						break
-					}
-					if len(data) > l {
-						data = data[l:]
-					}
+				}
+				if len(data) > l {
+					data = data[l:]
 				}
 			}
 		}
+
 	}
 }
 
-func (a *agent) disconnect_error(err error) {
+func (a *agent) reconnect_error(err error) {
 	if a.conn != nil {
 		err = &WorkerDisconnectError{
 			err:   err,
 			agent: a,
 		}
 		a.worker.err(err)
 	}
+	a.Connect()
 }
 
 func (a *agent) Close() {
+	if a.conn == nil {
+		return
+	}
 	a.Lock()
 	defer a.Unlock()
 	if a.conn != nil {
-		a.conn.Close()
+		_ = a.conn.Close()
 		a.conn = nil
 	}
 }
 
-func (a *agent) Grab() {
-	a.Lock()
-	defer a.Unlock()
-	a.grab()
+func (a *agent) Grab() (err error) {
+	if a.conn == nil {
+		return errors.New("No connection")
+	}
+	return a.grab()
 }
 
-func (a *agent) grab() {
+func (a *agent) grab() (err error) {
 	outpack := getOutPack()
 	outpack.dataType = rt.PT_GrabJobUniq
-	a.write(outpack)
+	return a.Write(outpack)
 }
 
-func (a *agent) PreSleep() {
-	a.Lock()
-	defer a.Unlock()
+func (a *agent) PreSleep() (err error) {
+	if a.conn == nil {
+		return errors.New("No connection")
+	}
 	outpack := getOutPack()
 	outpack.dataType = rt.PT_PreSleep
-	a.write(outpack)
+	return a.Write(outpack)
 }
 
-func (a *agent) reconnect() error {
-	a.Lock()
-	defer a.Unlock()
-	conn, err := net.Dial(a.net, a.addr)
-	if err != nil {
-		return err
+func (a *agent) grabReconnectState() (success bool) {
```

Comment:
I also think this should use `atomic.CompareAndSwapUint32()` instead of this double-lock/compare stuff

Reply:
implemented `lockReconnect()` like in client